var/spack/repos/builtin/packages/llvm/package.py

```diff
@@ -104,7 +104,7 @@ class Llvm(CMakePackage):
     depends_on('perl-data-dumper', type=('build'))
 
     # ncurses dependency
-    depends_on('ncurses+termlib')
+    depends_on('ncurses')
```

Comment:
This should be left as-is. LLVM will link against libtinfo and so it needs `ncurses+termlib`.

Reply:
I made termlib to be True by default in the ncurses/package.py so you'll get what you need, but also it will go around the concretizer bug that does not know to propagate specs to the whole dependency tree. In this case, the conflicting specs were ncurses+termlib and ncurses~termlib. That should have become ncurses+termlib.

Reply:
There is no concretizer bug with regard to ncurses. The +termlib variant of ncurses provides libtinfo, which llvm will link. The dependency spec for llvm is correct and the spec conflict is correct when ncurses~termlib is in the spec. The issue as I see it is what is the most sensible default value for the ncurses termlib variant? My opinion is that it should be `True`. If the ncurses termlib variant defaults to `True` there is nothing that would need to be done to modify packages that depend on ncurses, whether they specify `+termlib` or not.

Think of it this way, if the ncurses termlib variant defaults to `True`, someone could still set ncurses~termlib in their concretization preferences. LLVM would still need to have `depends_on('ncurses+termlib') specified to ensure that a spack built libtinfo is available. It does not hurt anything to have a package specify a +variant from a dependency even if the variant defaults to `True`.

Reply:
To be clear, I think what you are referring to by the concretizer bug is actually the limitations of the current concretizer behavior. That will changing soon so it will be interesting to see how these types of conflicts are resolved.

Reply:
I see. Sure, I can leave that +termlib there.

Reply:
Done.

Reply:
I have been thinking about the ncurses termlib variant and realized that it is more complicated than what I previously indicated. The termlib variant will use the `--with-termlib` configure option. That will build a separate tinfo library. Without that the symbols for tinfo will be in the ncurses library. I looked through spack packages that depend on ncurses. Some packages will use either library that it finds the symbols in but other packages require the library containing the tinfo symbols to be specified explicitly. 

That means that it is not simply a matter of changing the default value for the variant as all of the packages that depend on ncurses would need to be checked. As it is there are some packages, such as vim, that are broken if `ncurses+termlib` is in the spec.

Then there are the packages that require a separate tinfo library. I wonder if some of those could be made to use an ncurses library with the tinfo symbols. What happens in these cases is that a spack built ncurses library will be linked but the system tinfo library will be linked as well.

Reply:
Why is ncurses needed for llvm?

Reply:
Maybe we can compile a simpler llvm without ncurses?

Reply:
It needs it because the LLVM process wants to link against tinfo. From the build log, **without** the ncurses dependency in the package recipe:
```
...
/opt/packages/gpjohnsn/spack/lib/spack/env/gcc/g++  -std=c++11 -fPIC 
-fvisibility-inlines-hidden -Werror=date-time -std=c++11 -Wall -Wextra 
-Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers 
-pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-maybe-uninitialized 
-Wno-class-memaccess -Wno-noexcept-type -Wdelete-non-virtual-dtor 
-Wno-comment -ffunction-sections -fdata-sections -O3 -DNDEBUG   
-Wl,-allow-shlib-undefined  -Wl,-O3 -Wl,--gc-sections CMakeFiles/FileCheck.dir/FileCheck.cpp.o  -o ../../bin/FileCheck  -Wl,-rpath,/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-9.1.0/llvm-9.0.1-pfzajz3dn4ch3vtafj5pexq5r3ha4mjg/lib:/opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-9.1.0/llvm-9.0.1-pfzajz3dn4ch3vtafj5pexq5r3ha4mjg/lib64 -lpthread 
../../lib/libLLVMSupport.a -lz -lrt -ldl -ltinfo -lpthread -lm ../../lib/libLLVMDemangle.a
...
```

Reply:
> there are some packages, such as vim, that are broken if ncurses+termlib is in the spec.

If a package requires a variant to be specified in some way, it should generally be explicit about it. So for example the `vim` package should say `depends_on(ncurses~termlib)`.

> Then there are the packages that require a separate tinfo library. I wonder if some of those could be made to use an ncurses library with the tinfo symbols. What happens in these cases is that a spack built ncurses library will be linked but the system tinfo library will be linked as well.

Are you saying this occurs even when the package depends on `ncurses+termlib`?

Reply:
> If a package requires a variant to be specified in some way, it should generally be explicit about it.

Moreover, to be clear, I don't think this PR needs to be constrained to address each package that ought to explicitly specify `~termlib` before it gets merged. Although in the case of `vim` for example if we know up front that would be great to add.

Reply:
Would turning off this cmake variable turn this dependency off? LLVM_ENABLE_TERMINFO

Reply:
@glennpj apologies but could you highlight the specific part of https://github.com/spack/spack/pull/14795#discussion_r376562321 that is problematic?

Reply:
@danlipsa 

> Would turning off this cmake variable turn this dependency off? LLVM_ENABLE_TERMINFO

That is worth a try. I will take a look at that.

Reply:
@scheibelp For vim, there is:
```
configure_args.append("--with-tlib=ncursesw")
```
That would need to have a test added for +termlib and set to 'tinfow'. Something similar was done for samtools in #14507. I just picked out vim from a scan as I was looking into this.

Reply:
@danlipsa I think you found the variable we need. I want to do one more build with a different variant and then I will create a PR.

Reply:
Great! I'll remove the ncurses related changes from the PR and leave only the libxml2 stuff.